File: lang/c++/test/fun/vac/algorithms/meta/SortingTest.cpp
#include <fun/vac/algorithms/meta/Sorting.hpp>
#include <fun/vac/util/SequenceBuilder.hpp>
#include <fun/vac/util/Sequences.hpp>
#include <fun/vac/util/TestRunner.hpp>

using namespace std;

bool testBubbleSort() {
    size_t size = 32768;

    vector<int> arr (size);
    SequenceBuilder::packRandom(arr);

    int checksum = Sequences::parityChecksum(arr);

    Sorting::bubbleSort(arr);

    if (Sequences::parityChecksum(arr) != checksum) {
        return false;
    }

    if (!Sequences::isSorted(arr)) {
        return false;
    }

    return true;
}

bool testInsertionSort() {
    size_t size = 32768;

    vector<int> arr (size);
    SequenceBuilder::packRandom(arr);

    int checksum = Sequences::parityChecksum(arr);

    Sorting::insertionSort(arr);

    if (Sequences::parityChecksum(arr) != checksum) {
        return false;
    }

    if (!Sequences::isSorted(arr)) {
        return false;
    }

    return true;
}

bool testMergeSort() {
    size_t size = 32768;

    vector<int> arr (size);
    SequenceBuilder::packRandom(arr);

    int checksum = Sequences::parityChecksum(arr);

    Sorting::mergeSort(arr);

    if (Sequences::parityChecksum(arr) != checksum) {
        return false;
    }

    if (!Sequences::isSorted(arr)) {
        return false;
    }

    return true;
}

bool testQuickSort() {
    size_t size = 32768;

    vector<int> arr (size);
    SequenceBuilder::packRandom(arr);

    int checksum = Sequences::parityChecksum(arr);

    Sorting::quickSort(arr);

    if (Sequences::parityChecksum(arr) != checksum) {
        return false;
    }

    if (!Sequences::isSorted(arr)) {
        return false;
    }

    return true;
}

bool testSelectionSort() {
    size_t size = 32768;

    vector<int> arr (size);
    SequenceBuilder::packRandom(arr);

    int checksum = Sequences::parityChecksum(arr);

    Sorting::selectionSort(arr);

    if (Sequences::parityChecksum(arr) != checksum) {
        return false;
    }

    if (!Sequences::isSorted(arr)) {
        return false;
    }

    return true;
}

int main() {

    TestRunner::parseTest(testBubbleSort());

    TestRunner::parseTest(testInsertionSort());

    TestRunner::parseTest(testMergeSort());

    TestRunner::parseTest(testQuickSort());

    TestRunner::parseTest(testSelectionSort());
}

File: lang/ruby/test/fun/vac/algorithms/meta/SortingTest.rb
require "fun/vac/algorithms/meta/Sorting"
require "fun/vac/util/SequenceBuilder"
require "fun/vac/util/Sequences"
require "fun/vac/util/TestRunner"

def testBubbleSort()
    size = 32768

    arr = Array.new(size)
    SequenceBuilder.packRandom(arr)

    checksum = Sequences.parityChecksum(arr)

    Sorting.bubbleSort(arr)

    if Sequences.parityChecksum(arr) != checksum
        return false
    end

    unless Sequences.isSorted(arr)
        return false
    end

    return true
end

def testInsertionSort()
    size = 32768

    arr = Array.new(size)
    SequenceBuilder.packRandom(arr)

    checksum = Sequences.parityChecksum(arr)

    Sorting.insertionSort(arr)

    if Sequences.parityChecksum(arr) != checksum
        return false
    end

    unless Sequences.isSorted(arr)
        return false
    end

    return true
end

def testMergeSort()
    size = 32768

    arr = Array.new(size)
    SequenceBuilder.packRandom(arr)

    checksum = Sequences.parityChecksum(arr)

    Sorting.mergeSort(arr)

    if Sequences.parityChecksum(arr) != checksum
        return false
    end

    unless Sequences.isSorted(arr)
        return false
    end

    return true
end

def testQuickSort()
    size = 32768

    arr = Array.new(size)
    SequenceBuilder.packRandom(arr)

    checksum = Sequences.parityChecksum(arr)

    Sorting.quickSort(arr)

    if Sequences.parityChecksum(arr) != checksum
        return false
    end

    unless Sequences.isSorted(arr)
        return false
    end

    return true
end

def testSelectionSort()
    size = 32768

    arr = Array.new(size)
    SequenceBuilder.packRandom(arr)

    checksum = Sequences.parityChecksum(arr)

    Sorting.selectionSort(arr)

    if Sequences.parityChecksum(arr) != checksum
        return false
    end

    unless Sequences.isSorted(arr)
        return false
    end

    return true
end

if __FILE__ == $0

    TestRunner.parseTest(testBubbleSort())

    TestRunner.parseTest(testInsertionSort())

    TestRunner.parseTest(testMergeSort())

    TestRunner.parseTest(testQuickSort())

    TestRunner.parseTest(testSelectionSort())
end
